source/1_engine/rex_engine/src/cmdline.cpp

```diff
@@ -0,0 +1,205 @@
+
+namespace rex
+{
+  namespace cmdline
+  {
+    struct ActiveArgument
+    {
+      rsl::string_view argument;
+      rsl::string_view value;
+    };
+
+    class CommandLineArguments
+    {
+    public:
+      explicit CommandLineArguments(rsl::string_view cmdLine)
+      {
+        // verify the auto generated command line arguments
+        REX_ASSERT_X(verify_args(g_command_line_args.data(), g_command_line_args.size()), "You have ambuguous command line arguments");
+
+        if(cmdLine.empty())
+        {
+          REX_LOG(LogEngine, "No command line arguments to parse");
+          return;
+        }
+
+        parse_cmd_line(rsl::string_view(cmdLine));
+      }
+
+      rsl::optional<rsl::string_view> get_argument(rsl::string_view arg)
+      {
+        for(ActiveArgument active_arg: m_arguments)
+        {
+          // early optimization that strincmp can't do
+          if(arg.length() != active_arg.argument.length())
+            continue;
+
+          if(rsl::strincmp(arg.data(), active_arg.argument.data(), arg.length()) == 0)
+          {
+            return active_arg.value;
+          }
+        }
+
+        return rsl::nullopt;
+      }
+
+    private:
+      void parse_cmd_line(rsl::string_view cmdLine)
+      {
+        // all we have to do is split the command line based with spaces
+        // there's no need to perform any heap allocation for that.
+        count_t start_pos = cmdLine.find_first_not_of(' ', 0); // it's possible, although unlikely, that we start with a space
+
+        // all executables should be called with at least 1 argument
+        // the first argument should always be the path to the executable itself
+        REX_ASSERT_X(start_pos != -1, "No arguments given to your exe, there should always be at least 1 arg.");
+
+        // the module we're launching is always the first argument on the commandline
+        // we skip this
+        count_t space_pos = cmdLine.find_first_of(' ', start_pos);
+
+        // if we have no arguments, we just return
+        // this is possible when running without a debugger
+        if (space_pos == -1)
+        {
+          return;
+        }
+
+        start_pos         = space_pos + 1;
+
+        // get the new space pos
+        space_pos = cmdLine.find_first_of(' ', start_pos);
+
+        const rsl::string_view arg_prefix = "-"; // all arguments should start with a '-'
+        while(start_pos != -1 && space_pos != -1)
+        {
+          const count_t length            = space_pos - start_pos - arg_prefix.size();
+          const rsl::string_view argument = cmdLine.substr(start_pos + arg_prefix.size(), length);
+          add_argument(argument);
+          start_pos = cmdLine.find_first_not_of(' ', space_pos); // skip all additional spaces
+          space_pos = cmdLine.find_first_of(' ', start_pos);
+        }
+
+        if(start_pos != -1)
+        {
+          add_argument(cmdLine.substr(start_pos + arg_prefix.size())); // + 1 to ignore '-'
+        }
```

Comment:
Dunno if I asked this already or though about it. But what with multiple values for a single key.
eg: -Config=debug+release

Reply:
that's fine, in the logger_config.cpp you see how that's processed
arguments are seperated by spaces, in the example above, the value of the argument would be debug+release, the client is responsible for parsing that

Reply:
Perhaps some utility class might be created to extract values giving in a custom separator. Not for this PR though